### ISSUE TITLE
Parallelise BuildPackages.sh

### ISSUE THREAD

**Issue by stevelinton** [gap-system/gap#3624]
Please use the following template to submit an issue
(you may delete lines which are not used). Thank You!

### Observed behaviour

BuildPackages takes ages, even on a machine with fast disks and lots of cores.

### Expected behaviour

It would be nice if it was quicker. It could use make -j on packages for which ithat works, and/or build multiple packages in parallel.

**Comment by fingolfin** [comment]
(General caveat: I am not sure how far I want to go in extending this script which always was meant as an interim hack until we have something better. On the other hand, it's not clear we ever will get something better. Well, perhaps https://github.com/gap-packages/PackageManager will get there eventually).

A simple way to get this to happen is to use the `MAKEFLAGS` environment variable, like so: `MAKEFLAGS=-j8 ../bin/BuildPackages.sh`. But that will "only" parallelize within a single build (and only if it uses `make`). Another orthogonal approach is to indeed build multiple packages in parallel. On systems with `xargs` installed (so for most Linux distros and macOS), one can do e.g. this (for 8 parallel jobs):  `ls | xargs -n1 -P8 ../bin/BuildPackages.sh` (this is a bit of hack, though, instead of `ls` it would be better to generate a list of package directory names; but it works OK enough). The big advantage of that is that it also parallelizes `configure` runs, which can take up some time (this effect is much more prominent on macOS, and even more so on Windows, due to their very slow process startup times, which dominates the performance of GNU `configure` scripts).

Here are some timings from an somewhat older Ubuntu quad-core server (i7-3820 CPU @ 3.60GHz). For a quadcore, usually 8 jobs is a good compromise that prevents IO and/or memory saturation, for other machines one may of course want to use different parameters.

Without any parallelization:
```
$ time ../bin/BuildPackages.sh
real	6m32.992s
user	5m23.264s
sys	0m12.540s
```

With `MAKEFLAGS=-j8`
```
$ time MAKEFLAGS=-j8 ../bin/BuildPackages.sh
real	4m40.908s
user	5m12.872s
sys	0m11.308s
```

With `xargs -P8`:
```
$ time ls | xargs -n1 -P8 ../bin/BuildPackages.sh
real	3m28.195s
user	5m16.572s
sys	0m12.852s
```

Of course one can also combine the two techniques, but I'd recommend to not oversaturate the computer But e.g. using `MAKEFLAGS=-j3` and `xargs -P3` might be a good compromise. Let's try that
```
$ time ls | MAKEFLAGS=-j3 xargs -n1 -P3 ../bin/BuildPackages.sh
real	3m12.468s
user	5m1.800s
sys	0m10.516s
```

Changing the numbers doesn't matter much at this point:
```
$ time ls | MAKEFLAGS=-j4 xargs -n1 -P4 ../bin/BuildPackages.sh
real	3m11.264s
user	5m8.876s
sys	0m11.384s
```

But of course, this isn't flexible if one job is stuck building a single big package. Indeed, this happens for me: I noticed that actually a substantial part of the time was spent in compiling `Normaliz` for the `NormalizInterface` package, which I maintain; it contains some heavy C++ template code, which takes forever to compile; this could be mitigate by removing that package. Here are the numbers without `NormalizInterface`:
```
$ time ../bin/BuildPackages.sh

$ time MAKEFLAGS=-j8 ../bin/BuildPackages.sh
real	2m1.977s
user	2m16.912s
sys	0m7.228s

$ time ls | xargs -n1 -P8 ../bin/BuildPackages.sh
real	1m20.254s
user	2m30.496s
sys	0m8.744s

$ time ls | MAKEFLAGS=-j4 xargs -n1 -P4 ../bin/BuildPackages.sh
real	0m52.881s
user	2m19.772s
sys	0m7.688s
```

One cheap trick to get better load balancing is to abuse the fact that `make` (well, at least `GNU make`, but probably also other make versions?) performs load balancing between any sub-instances of `make` that are launched by the parent `make` process. So, one could generate a temporary `Makefile` in the `pkg` dir, which calls the build script for every package in separate targets, and then simply do `make -j8` to get "optimal" parallelization -- well at least in theory. I did not actually try it.


All of this is of course not something I'd want to suggest in general, but (a) it is a quick way to get this faster *now* for you, and (b) it might give some ideas for enhancing `BuildPackages`. But note that in general, "guessing" a good number of parallel jobs is not possible in a portable fashion (though I guess we could try to at least do it for macOS and Linux); so there probably should be a `--max-jobs=N` argument to `BuildPackages.sh`
